File: LeetCode/Easy/0141_Linked_List_Cycle/Solution.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
	single, double := head, head.Next
	for single != nil && double != nil {
		if single == double || single == double.Next {
			return true
		}

		single = single.Next
		double = double.Next
		if double != nil {
			double = double.Next
		}
	}
	return false
}

File: LeetCode/Easy/0141_Linked_List_Cycle/Solution.java

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

//2021-11-16
public class Solution {
    public boolean hasCycle(ListNode head) {
        if (head == null) return false;
        Set<ListNode> seen = new HashSet<ListNode>();
        return isThereCycle(head, seen);
    }
    
    private boolean isThereCycle(ListNode head, Set<ListNode> seen) {
        if (head == null) return false;
        if (seen.contains(head)) return true;
        seen.add(head);
        return isThereCycle(head.next, seen);
    }
}